src/rdb/rdb_raft.c

```diff
@@ -2430,57 +2445,95 @@ rdb_raft_get_ae_max_size(void)
 	return value;
 }
 
-/*
- * Load raft persistent state, if any. Our raft callbacks must be registered
- * already, because rdb_raft_cb_notify_membership_event is required. We use
- * db->d_raft_loaded to instruct some of our raft callbacks to avoid
- * unnecessary write I/Os.
- */
-static int
-rdb_raft_load(struct rdb *db)
+int
+rdb_raft_dictate(struct rdb *db)
 {
-	d_iov_t		value;
-	uint64_t	term;
-	int		vote;
-	int		rc;
+	struct rdb_lc_record	lc_record = db->d_lc_record;
+	uint64_t		term;
+	d_rank_list_t		replicas;
+	d_rank_t		self = dss_self_rank();
+	d_iov_t			keys[2];
+	d_iov_t			value;
+	uint64_t		index = lc_record.dlr_tail;
+	int			rc;
 
-	D_DEBUG(DB_MD, DF_DB": load persistent state: begin\n", DP_DB(db));
-	D_ASSERT(!db->d_raft_loaded);
+	/*
+	 * Since we don't have an RDB fsck phase yet, do a basic check to avoid
+	 * arithmetic issues.
+	 */
+	if (lc_record.dlr_base >= index) {
+		D_ERROR(DF_DB": LC record corrupted: base "DF_U64" >= tail "DF_U64"\n", DP_DB(db),
+			lc_record.dlr_base, index);
+		return -DER_IO;
+	}
 
-	d_iov_set(&value, &term, sizeof(term));
-	rc = rdb_mc_lookup(db->d_mc, RDB_MC_ATTRS, &rdb_mc_term, &value);
-	if (rc == 0) {
-		rc = raft_set_current_term(db->d_raft, term);
-		D_ASSERTF(rc == 0, DF_RC"\n", DP_RC(rc));
-	} else if (rc != -DER_NONEXIST) {
-		goto out;
+	/* Get the term at the last index. */
+	if (index - lc_record.dlr_base - 1 > 0) {
+		struct rdb_entry header;
+
+		/* The LC has entries. Get from the last entry. */
+		d_iov_set(&value, &header, sizeof(header));
+		rc = rdb_lc_lookup(db->d_lc, index - 1, RDB_LC_ATTRS, &rdb_lc_entry_header, &value);
+		if (rc != 0) {
+			D_ERROR(DF_DB": failed to look up entry "DF_U64" header: "DF_RC"\n",
+				DP_DB(db), index - 1, DP_RC(rc));
+			return rc;
+		}
+		term = header.dre_term;
+	} else {
+		/* The LC has no entries. Get from the snapshot. */
+		term = lc_record.dlr_base_term;
 	}
 
-	d_iov_set(&value, &vote, sizeof(vote));
-	rc = rdb_mc_lookup(db->d_mc, RDB_MC_ATTRS, &rdb_mc_vote, &value);
-	if (rc == 0) {
-		rc = raft_vote_for_nodeid(db->d_raft, vote);
-		D_ASSERTF(rc == 0, DF_RC"\n", DP_RC(rc));
-	} else if (rc != -DER_NONEXIST) {
-		goto out;
+	/*
+	 * At a new index, reset the membership to only ourself. We also punch
+	 * the entry header and data just for consistency, for this may be a
+	 * membership change entry that, for instance, adds a node other than
+	 * ourself, which contradicts with the new membership of only ourself.
+ 	 */
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs